src/main/java/com/redhat/pantheon/sling/ServiceResourceResolverProvider.java

```diff
@@ -0,0 +1,38 @@
+package com.redhat.pantheon.sling;
+
+import org.apache.sling.api.resource.LoginException;
+import org.apache.sling.api.resource.ResourceResolver;
+import org.apache.sling.api.resource.ResourceResolverFactory;
+import org.osgi.service.component.annotations.Activate;
+import org.osgi.service.component.annotations.Component;
+import org.osgi.service.component.annotations.Reference;
+
+import java.util.function.Consumer;
+
+@Component(
+        service = ServiceResourceResolverProvider.class
+)
+public class ServiceResourceResolverProvider {
+
+    private ResourceResolverFactory resourceResolverFactory;
+
+    @Activate
+    public ServiceResourceResolverProvider(
+            @Reference ResourceResolverFactory resourceResolverFactory) {
+        this.resourceResolverFactory = resourceResolverFactory;
+    }
+
+    public ResourceResolver getServiceResourceResolver() {
+        try {
+            return resourceResolverFactory.getServiceResourceResolver(null);
+        } catch (LoginException e) {
+            throw new RuntimeException(e);
+        }
+    }
+
+    public void withServiceResourceResolver(Consumer<ResourceResolver> function) {
+        final ResourceResolver serviceResourceResolver = this.getServiceResourceResolver();
+        function.accept(serviceResourceResolver);
+        serviceResourceResolver.close();
+    }
+}
```

Comment:
it's all in the resource resolver. As long as you read/write the resources with the admin-level resolver you get the different privileges.  That's partly what I liked about my functional method above, the use of the resolver is constrained to the lambda. The render servlet invocation is still done as the logged in user, it's just the additional resolver gives the code different privileges.

We can talk a bit more about how the `getServiceResourceResolver` method works. There's some minor depth to it, but this is basically it, the code just asks for a service resource resolver, and as long as the system is well configured, you get one.

Reply:
That is very cool. I'm very curious how you figured that out, because this approach is not at all obvious when I read through the article that you emailed me awhile back: https://sling.apache.org/documentation/the-sling-engine/service-authentication.html

Say, I wonder if we can use this new service user approach in our `SlingRepositoryInitializer` class rather than the deprecated call to `loginAdministrative()` that we're currently using?